components/eamxx/src/share/io/scream_scorpio_types.hpp

```diff
@@ -0,0 +1,154 @@
+#ifndef SCREAM_SCORPIO_TYPES_HPP
+#define SCREAM_SCORPIO_TYPES_HPP
+
+#include <string>
+#include <map>
+#include <vector>
+#include <memory>
+
+namespace scream {
+namespace scorpio {
+
+// Enum denoting how we open a file
+enum FileMode {
+  Unset = 0,
+  Read = 1,
+  Write = 2,
+  Append = Read | Write
+};
+std::string e2str (const FileMode mode);
+
+// I/O types supported
+enum IOType {
+  // Default I/O type is used to let the code choose I/O type as needed (via CIME)
+  DefaultIOType = 0,
+  NetCDF,
+  PnetCDF,
+  Adios,
+  Hdf5,
+  Invalid
+};
+
+IOType str2iotype(const std::string &str);
+std::string iotype2str(const IOType iotype);
+
+// The type used by PIOc for offsets
+using offset_t = std::int64_t;
+
+/*
+ * The following PIOxyz types each represent an entity that
+ * is associated in scorpio with an id. For each of them, we add some
+ * additional metadata, to avoid having to call PIOc_inq_xyz every time
+ * we need such information.
+ * While these types are defined in this public header, the customers
+ * of scream_io will never be able to get any object of these types out
+ * of the internal database. In particular, all data is stored in a
+ * ScorpioSession singleton class, whose declaration is hidden inside
+ * scream_scorpio_interface.cpp.
+ */
+
+// The basic common data of any PIO entity
+struct PIOEntity {
+  int ncid = -1;      // PIO access all data via their id
+  std::string name;   // In EAMxx, we prefer to use a string name than a number id
+};
+
+// An entity that is linked to a specific file
+// This allows an entity to retrieve its file
+struct PIOFileEntity : public PIOEntity {
+  int fid = -1;       // Allow retrieving file from the entity
+};
+
+// struct PIOAtt : public PIOFileEntity {
+//   int       varid = -1;
+//   ekat::any val;
+// };
+
+// A dimension
+struct PIODim : public PIOFileEntity {
+  int length       = -1;
+  bool unlimited   = false;
+
+  // In case we decompose the dimension, this will store
+  // the owned offsets on this rank
+  // NOTE: use a pointer, so we can detect if a decomposition already
+  //       existed or not when we set one.
+  std::shared_ptr<std::vector<offset_t>> offsets;
+};
+
+// A decomposition
+// NOTE: the offsets of a PIODecomp are not the same as the offsets of
+//       stored in its dim. The latter are the offsets *along that dim*.
+//       A PIODecomp is associated with a Nd layout, which includes decomp_dim
+//       among its dimensions. PIODecomp::offsets are the offsets of the full
+//       array layout owned by this rank. Hence, there can be many PIODecomp
+//       all storing the same dim
+struct PIODecomp : public PIOEntity {
+  std::vector<offset_t>           offsets;  // Owned offsets
+  std::shared_ptr<const PIODim>   dim; 
+};
+
+// // A decomposition of a single dimension
```

Comment:
Any reason the next 10 lines are commented out?

Reply:
Yeah, these were early iterations, which I forgot to clean up.. will clean it up once everything passes.